staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go

```diff
@@ -119,36 +120,59 @@ func newDefaultFieldManager(f Manager, typeConverter internal.TypeConverter, obj
 	f = NewLastAppliedManager(f, typeConverter, objectConverter, kind.GroupVersion())
 	f = NewLastAppliedUpdater(f)
 
-	return NewFieldManager(f)
+	return NewFieldManager(f, isSubresource)
 }
 
-// Update is used when the object has already been merged (non-apply
-// use-case), and simply updates the managed fields in the output
-// object.
-func (f *FieldManager) Update(liveObj, newObj runtime.Object, manager string) (object runtime.Object, err error) {
+func decodeLiveManagedFields(liveObj runtime.Object) (Managed, error) {
+	liveAccessor, err := meta.Accessor(liveObj)
+	if err != nil {
+		return nil, err
+	}
+	managed, err := internal.DecodeObjectManagedFields(liveAccessor.GetManagedFields())
+	if err != nil {
+		return internal.NewEmptyManaged(), nil
+	}
+	return managed, nil
+}
+
+func decodeManagedFields(liveObj, newObj runtime.Object, isSubresource bool) (Managed, error) {
 	// If the object doesn't have metadata, we should just return without trying to
 	// set the managedFields at all, so creates/updates/patches will work normally.
 	newAccessor, err := meta.Accessor(newObj)
 	if err != nil {
-		return newObj, nil
+		return nil, err
+	}
+
+	if isResetManagedFields(newAccessor.GetManagedFields()) {
+		return internal.NewEmptyManaged(), nil
 	}
 
+	// We take the managedFields of the live object in case the request tries to
+	// manually set managedFields via a subresource.
+	if isSubresource {
```

Comment:
this should come before the "isResetManagedFields" method above. Otherwise, someone would be able to reset the managed fields through the subresource, which we probably don't want.

Reply:
Totally, thanks for catching it!

Reply:
Done in c8b4842, I have also added a unit test to make sure this case is covered.

Reply:
This check can be done first in this function, no? The accessor is not used if this is for a subresource?

Reply:
Agree, I think the subresource check should go first. Thanks for spotting it!